File: app/src/main/java/com/qimian233/ztool/hook/modules/wallpaper/ChargeAnimationFixModule.java
package com.qimian233.ztool.hook.modules.wallpaper;

import com.qimian233.ztool.data.ScopeKeys;
import com.qimian233.ztool.hook.base.AppHookModule;

import io.github.libxposed.api.XposedInterface;
import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Method;

/**
 * 充电动画修复模块
 * 修复ZUI系统壁纸设置中的充电动画显示问题，强制显示全部充电动画选项
 * 通过修改Utilities类的关键方法，确保系统使用包含全部充电动画的资源数组
 */
public class ChargeAnimationFixModule extends AppHookModule {
    private static final String UTILS_CLASS = "com.zui.wallpapersetting.util.Utilities";

    public ChargeAnimationFixModule() {}

    @Override
    public String getModuleName() {
        return "charge_animation_fix";  // 模块唯一标识
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                ScopeKeys.WALLPAPER_SETTINGS.packageName  // ZUI壁纸设置应用
        };
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if (!ScopeKeys.WALLPAPER_SETTINGS.packageName.equals(packageName)) {
            return;  // 提前返回，避免不必要的处理
        }

        // 检查模块是否已启用
        if (!isEnabled()) {
            logger.info("Module is disabled, skipping hook");
            return;
        }

        try {
            hookChargeAnimationUtils(classLoader);
        } catch (Throwable t) {
            logger.error("Failed to hook charge animation utilities", t);
        }
    }

    /**
     * Hook Utilities类的关键方法，修复充电动画显示
     */
    private void hookChargeAnimationUtils(ClassLoader classLoader) {
        try {
            Class<?> utilsClass = classLoader.loadClass(UTILS_CLASS);

            // 修改Utilities.isLegiony()返回true
            // 原逻辑：(!Utilities.isLegiony() || Utilities.isOversea) ? "chargeStyle_row" : "chargeStyle"
            // 通过强制isLegiony返回true，确保使用"chargeStyle"数组
            Method isLegionyMethod = utilsClass.getDeclaredMethod("isLegiony");
            hookWithId(isLegionyMethod, "is_legiony", chain -> true);

            // 修改Utilities.isOversea()返回false
            Method isOverseaMethod = utilsClass.getDeclaredMethod("isOversea");
            hookWithId(isOverseaMethod, "is_oversea", chain -> false);

            // 修复平板设备的充电动画显示问题
            Method isPadMethod = utilsClass.getDeclaredMethod("isPad");
            hookWithId(isPadMethod, "is_pad", chain -> false);

            logger.info("Successfully enabled all charge animations");
            logger.debug("Now showing: default, particle, turbo, triangle, girl");
        } catch (Throwable t) {
            logger.error("Failed to hook Utilities class", t);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/wallpaper/ChargeAnimationFixModule.kt
package com.qimian233.ztool.hook.modules.wallpaper

import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam

/**
 * Charge animation fix module.
 * Fixes the charge animation display issue in the ZUI system wallpaper settings,
 * forcing all charge animation options to be shown.
 * By modifying key methods of the Utilities class, ensures the system uses the
 * resource array containing all charge animations.
 */
class ChargeAnimationFixModule : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.CHARGE_ANIMATION_FIX.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.WALLPAPER_SETTINGS.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        try {
            hookChargeAnimationUtils(classLoader)
        } catch (t: Throwable) {
            logger.error("Failed to hook charge animation utilities", t)
        }
    }

    /**
     * Hooks key methods of the Utilities class to fix charge animation display.
     */
    private fun hookChargeAnimationUtils(classLoader: ClassLoader) {
        try {
            val utilsClass = classLoader.loadClass(UTILS_CLASS)

            // Force Utilities.isLegiony() to return true
            // Original logic: (!Utilities.isLegiony() || Utilities.isOversea) ? "chargeStyle_row" : "chargeStyle"
            // Forcing isLegiony to return true ensures the "chargeStyle" array is used
            val isLegionyMethod = utilsClass.getDeclaredMethod("isLegiony")
            hookWithId(
                isLegionyMethod,
                "is_legiony"
            ) { true }

            // Force Utilities.isOversea() to return false
            val isOverseaMethod = utilsClass.getDeclaredMethod("isOversea")
            hookWithId(
                isOverseaMethod,
                "is_oversea"
            ) { false }

            // Fix charge animation display on tablet devices
            val isPadMethod = utilsClass.getDeclaredMethod("isPad")
            hookWithId(isPadMethod, "is_pad") { false }

            logger.info("Successfully enabled all charge animations")
            logger.debug("Now showing: default, particle, turbo, triangle, girl")
        } catch (t: Throwable) {
            logger.error("Failed to hook Utilities class", t)
        }
    }

    companion object {
        private const val UTILS_CLASS = "com.zui.wallpapersetting.util.Utilities"
    }
}
